File: source/ch/cyberduck/core/local/FoundationProgressIconService.java
package ch.cyberduck.core.local;

/*
 * Copyright (c) 2012 David Kocher. All rights reserved.
 * http://cyberduck.ch/
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * Bug fixes, suggestions and comments should be sent to:
 * dkocher@cyberduck.ch
 */

import ch.cyberduck.core.Local;
import ch.cyberduck.core.library.Native;
import ch.cyberduck.core.transfer.TransferStatus;
import ch.cyberduck.ui.cocoa.application.NSWorkspace;

import org.apache.log4j.Logger;

/**
 * @version $Id$
 */
public final class FoundationProgressIconService implements IconService {
    private static final Logger log = Logger.getLogger(FoundationProgressIconService.class);

    public static void register() {
        if(Factory.VERSION_PLATFORM.matches("10\\.8.*")) {
            IconServiceFactory.addFactory(Factory.NATIVE_PLATFORM, new Factory());
        }
        else {
            log.warn(String.format("Skip registering notifications on %s", Factory.VERSION_PLATFORM));
        }
    }

    private static class Factory extends IconServiceFactory {
        @Override
        protected IconService create() {
            return new FoundationProgressIconService();
        }
    }

    private FoundationProgressIconService() {
        //
    }

    static {
        Native.load("FoundationProgressService");
    }

    @Override
    public boolean set(final Local file, final String image) {
        return false;
    }

    @Override
    public boolean set(final Local file, final TransferStatus status) {
        this.progress(file.getAbsolute(), status.getCurrent(), status.getLength());
        synchronized(NSWorkspace.class) {
            final NSWorkspace workspace = NSWorkspace.sharedWorkspace();
            workspace.noteFileSystemChanged(file.getAbsolute());
        }
        return true;
    }

    @Override
    public boolean remove(final Local file) {
        this.cancel(file.getAbsolute());
        return true;
    }

    private native void progress(String file, long current, long size);

    private native void cancel(String file);
}

File: source/ch/cyberduck/core/local/FoundationProgressIconService.m
/*
 *  Copyright (c) 2012 David Kocher. All rights reserved.
 *  http://cyberduck.ch/
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  Bug fixes, suggestions and comments should be sent to:
 *  dkocher@cyberduck.ch
 */

#include <stdio.h>

#import <FoundationProgressIconService.h>
#import <Foundation/Foundation.h>
#import <JavaNativeFoundation/JNFString.h>

JNIEXPORT void JNICALL Java_ch_cyberduck_core_local_FoundationProgressIconService_progress(JNIEnv *env, jobject this, jstring file, jlong current, jlong size)
{
    if(NSClassFromString(@"NSProgress")) {
        NSDictionary* info = [NSDictionary dictionaryWithObjectsAndKeys:
            @"NSProgressFileOperationKindDownloading", @"NSProgressFileOperationKindKey",
            [NSURL fileURLWithPath:JNFJavaToNSString(env, file)], @"NSProgressFileURLKey",
            nil];
        id progress = [NSClassFromString(@"NSProgress") performSelector:@selector(currentProgress)];
        if(nil == progress) {
            progress = [NSClassFromString(@"NSProgress") performSelector:@selector(alloc)];
            [progress performSelector:@selector(initWithParent:userInfo:)
                          withObject:nil
                          withObject:info];
            [progress setTotalUnitCount:[[NSNumber numberWithFloat:size] integerValue]];
            [progress setCompletedUnitCount:[[NSNumber numberWithFloat:current] integerValue]];
            [progress setKind:@"NSProgressKindFile"];
            [progress setPausable:NO];
            [progress setCancellable:NO];
            // Sets the receiver as the current progress object of the current thread.
            [progress becomeCurrentWithPendingUnitCount:size];
            [progress publish];
        }
        else {
            [progress setTotalUnitCount:[[NSNumber numberWithFloat:size] integerValue]];
            [progress setCompletedUnitCount:[[NSNumber numberWithFloat:current] integerValue]];
        }
    }
}

JNIEXPORT void JNICALL Java_ch_cyberduck_core_local_FoundationProgressIconService_cancel(JNIEnv *env, jobject this, jstring file)
{
    if(NSClassFromString(@"NSProgress")) {
        id progress = [NSClassFromString(@"NSProgress") performSelector:@selector(currentProgress)];
        if(nil == progress) {
            return;
        }
        // Balance the most recent previous invocation of becomeCurrentWithPendingUnitCount: on the same thread
        // by restoring the current progress object
        [progress resignCurrent];
        // Invoke cancelation handler
        [progress cancel];
    }
}